api/python/quilt3/packages.py

```diff
@@ -851,20 +852,33 @@ def set_meta(self, meta):
         return self
 
     def _fix_sha256(self):
-        entries = [entry for key, entry in self.walk() if entry.hash is None]
-        if not entries:
-            return
+        """
+        Calculate and set missing hash values
+        """
+        def _fix_pkg_entries(entries):
+            physical_keys = []
+            sizes = []
+            for entry in entries:
+                physical_keys.append(entry.physical_key)
+                sizes.append(entry.size)
 
-        physical_keys = []
-        sizes = []
-        for entry in entries:
-            physical_keys.append(entry.physical_key)
-            sizes.append(entry.size)
+            results = calculate_sha256(physical_keys, sizes)
+
+            entries_w_missing_hash = []
+            for entry, obj_hash in zip(entries, results):
+                if obj_hash is None:
+                    entries_w_missing_hash.append(entry)
+                else:
+                    entry.hash = dict(type='SHA256', value=obj_hash) if obj_hash is not None else None
+            return entries_w_missing_hash
+
+        entries = [entry for key, entry in self.walk() if entry.hash is None]
 
-        results = calculate_sha256(physical_keys, sizes)
+        retries = 0
+        while entries and retries < MAX_FIX_HASH_RETRIES:
+            entries = _fix_pkg_entries(entries)
```

Comment:
We should use tenacity to backoff and retry here, otherwise we could needlessly hammer a just-failed call.